File: tests/st/ops/test_ops_lin_space_ext.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import numpy as np
import pytest
from tests.st.utils import test_utils
from tests.st.ops.dynamic_shape.test_op_utils import TEST_OP
from mindspore import ops
from mindspore.ops.function.math_func import linspace_ext
import mindspore as ms
from mindspore.common import mutable

def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)

@test_utils.run_with_cell
def lin_space_ext_forward_func(start, end, steps, dtype=None):
    return linspace_ext(start, end, steps, dtype=dtype)

@test_utils.run_with_cell
def lin_space_ext_backward_func(start, end, steps, dtype=None):
    return ops.grad(lin_space_ext_forward_func, (0, 1))(start, end, steps, dtype)


@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_arm_ascend_training
@pytest.mark.parametrize('mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@pytest.mark.parametrize('dtype', [ms.float32])
def test_lin_space_ext_normal(mode, dtype):
    """
    Feature: Ops.
    Description: test op LinSpaceExt forward and backward.
    Expectation: expect correct result.
    """
    ms.context.set_context(mode=mode)
    ms.set_context(jit_level='O0')
    ## forward
    start_scalar, end_scalar, steps_scalar = 5, 25.13137, 3
    output1 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect1 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output1.asnumpy(), expect1)

    start_scalar, end_scalar, steps_scalar = 1.0241, 25.0, 7
    dtype = ms.float32
    output3 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect3 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output3.asnumpy(), expect3)

    start_scalar, end_scalar, steps_scalar = -5.0123, 250, 14
    dtype = ms.float32
    output5 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect5 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output5.asnumpy(), expect5)


@pytest.mark.level1
@pytest.mark.env_onecard
@pytest.mark.platform_arm_ascend910b_training
@pytest.mark.parametrize('mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@pytest.mark.parametrize('dtype', [ms.bfloat16])
def test_lin_space_ext_bfloat16(mode, dtype):
    """
    Feature: Ops.
    Description: test op LinSpaceExt.
    Expectation: expect correct result.
    """
    ms.context.set_context(mode=mode)
    ms.set_context(jit_level='O0')

    start_scalar, end_scalar, steps_scalar = 5, 25.5234, 7
    output1 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect1 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output1.float().asnumpy(), expect1, rtol=4e-3, atol=4e-3)

    start_scalar, end_scalar, steps_scalar = 1.0084, -25.045893, 23
    dtype = ms.float32
    output3 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect3 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output3.float().asnumpy(), expect3, rtol=4e-3, atol=4e-3)

    start_scalar, end_scalar, steps_scalar = -5.204134, 250.43108, 17
    dtype = ms.float32
    output5 = lin_space_ext_forward_func(start_scalar, end_scalar, steps_scalar, dtype)
    expect5 = np.linspace(start_scalar, end_scalar, steps_scalar, axis=-1)
    assert np.allclose(output5.float().asnumpy(), expect5, rtol=4e-3, atol=4e-3)


@pytest.mark.level1
@pytest.mark.env_onecard
@pytest.mark.platform_arm_ascend_training
def test_lin_space_ext_dynamic():
    """
    Feature: test dynamic by TEST_OP.
    Description: test op concat.
    Expectation: expect tile result.
    """
    input_case1 = (5.4113, -130.13134, 5)
    input_case2 = (-2421.13178, 413.47, 6004)
    TEST_OP(lin_space_ext_forward_func, [[*input_case1], [*input_case2]], '', disable_yaml_check=True,
            disable_input_check=True, disable_mode=['GRAPH_MODE'], disable_nontensor_dynamic_type='BOTH')

    input_case3 = (5, 50.23, mutable(5), ms.int32)
    input_case4 = (-5, 43.97, mutable(13), ms.float32)
    TEST_OP(lin_space_ext_forward_func, [[*input_case3], [*input_case4]], '', disable_yaml_check=True,
            disable_input_check=True, disable_mode=['GRAPH_MODE'], disable_nontensor_dynamic_type='BOTH',
            disable_resize=True)

File: tests/ut/cpp/ops/test_ops_lin_space_ext.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ops/test_ops.h"
#include "ops/ops_func_impl/lin_space_ext.h"
#include "ops/test_value_utils.h"
#include "ops/test_ops_cmp_utils.h"

namespace mindspore {
namespace ops {
struct LinSpaceExtParams {
  ValuePtr start;
  ValuePtr end;
  ValuePtr steps;
  ValuePtr dtype;
  ShapeVector output_shape;
  TypePtr output_type;
};

class TestLinSpaceExt : public TestOps, public testing::WithParamInterface<LinSpaceExtParams> {};
TEST_P(TestLinSpaceExt, dyn_shape) {
  const auto &param = GetParam();
  auto start = param.start->ToAbstract();
  auto end = param.end->ToAbstract();
  auto steps = param.steps->ToAbstract();
  auto dtype = param.dtype->ToAbstract();
  auto expect = std::make_shared<abstract::AbstractTensor>(param.output_type, param.output_shape);

  LinSpaceExtFuncImpl lin_space_ext_func_impl;
  auto prim = std::make_shared<Primitive>("LinSpaceExt");

  auto out_dtype = lin_space_ext_func_impl.InferType(prim, {start, end, steps, dtype});
  ASSERT_TRUE(*out_dtype == *expect->GetType());
  auto out_shape = lin_space_ext_func_impl.InferShape(prim, {start, end, steps, dtype});
  ASSERT_TRUE(*out_shape == *expect->GetShape());
}

INSTANTIATE_TEST_CASE_P(TestLinSpaceExt, TestLinSpaceExt,
                        testing::Values(LinSpaceExtParams{CreateScalar<int64_t>(3),
                                                          CreateScalar<int64_t>(10),
                                                          CreateScalar<int64_t>(3),
                                                          CreatePyInt(kNumberTypeFloat32),
                                                          {3},
                                                          kFloat32},
                                        LinSpaceExtParams{CreateScalar<double>(-31.414),
                                                          CreateScalar<double>(3413.54598),
                                                          CreateScalar<int64_t>(123),
                                                          CreatePyInt(kNumberTypeFloat32),
                                                          {123},
                                                          kFloat32},
                                        LinSpaceExtParams{CreateScalar<float>(3.0123),
                                                          CreateScalar<float>(-10.011),
                                                          CreateScalar(kValueAny),
                                                          CreatePyInt(kNumberTypeFloat64),
                                                          {-1},
                                                          kFloat64}));

class TestLinSpaceExtSimpleInfer : public TestOps, public testing::WithParamInterface<LinSpaceExtParams> {};
TEST_P(TestLinSpaceExtSimpleInfer, simple_infer) {
  const auto &param = GetParam();
  LinSpaceExtFuncImpl lin_space_ext_func_impl;
  auto prim = std::make_shared<Primitive>("LinSpaceExt");
  ASSERT_NE(prim, nullptr);
  ValuePtrList input_values;
  input_values.push_back(std::move(param.start));
  input_values.push_back(std::move(param.end));
  input_values.push_back(std::move(param.steps));
  input_values.push_back(std::move(param.dtype));

  auto expect_shape = ShapeArray{param.output_shape};
  auto expect_type = TypePtrList{param.output_type};

  auto output_shape = lin_space_ext_func_impl.InferShape(prim, input_values);
  auto output_type = lin_space_ext_func_impl.InferType(prim, input_values);

  ShapeCompare(output_shape, expect_shape);
  TypeCompare(output_type, expect_type);
}
INSTANTIATE_TEST_CASE_P(TestLinSpaceExtSimpleInfer, TestLinSpaceExtSimpleInfer,
                        testing::Values(LinSpaceExtParams{CreateScalar<int64_t>(3),
                                                          CreateScalar<int64_t>(10),
                                                          CreateScalar<int64_t>(3),
                                                          CreatePyInt(kNumberTypeFloat32),
                                                          {3},
                                                          kFloat32},
                                        LinSpaceExtParams{CreateScalar<double>(-31.414),
                                                          CreateScalar<double>(3413.54598),
                                                          CreateScalar<int64_t>(123),
                                                          CreatePyInt(kNumberTypeFloat32),
                                                          {123},
                                                          kFloat32},
                                        LinSpaceExtParams{CreateScalar<float>(3.0123),
                                                          CreateScalar<float>(-10.011),
                                                          CreateScalar(kValueAny),
                                                          CreatePyInt(kNumberTypeFloat64),
                                                          {-1},
                                                          kFloat64}));
}  // namespace ops
}  // namespace mindspore
